lib/posix/pthread_key.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <posix/pthread.h>
+
+static sem_t sem;
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *newKey;
+
+	*key = NULL;
+
+	sem_init(&sem, 0, 1);
+
+	newKey = (pthread_key_obj *)k_malloc(sizeof(pthread_key_obj));
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	sys_dlist_init(&(newKey->KeyDataL));
+
+	newKey->destructor = destructor;
+	*key = (void *)newKey;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	KeyData *keyData;
+
+	sem_wait(&sem);
+
+	/* Delete thread-specific elements associated with the key */
+	while (!sys_dlist_is_empty(&(keyObj->KeyDataL))) {
+
+		/* Remove the object from the list KeyDataL */
+		keyData = (KeyData *)sys_dlist_get(&(keyObj->KeyDataL));
+
+		/* Deallocate the object's memory */
+		k_free((void *)keyData);
+	}
+
+	k_free((void *)keyObj);
+
+	sem_post(&sem);
+
+	return 0;
+}
+
+/**
+ * @brief Associate a thread-specific value with a key
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setspecific(pthread_key_t key, const void *value)
+{
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	struct posix_thread *thread = (struct posix_thread *)pthread_self();
+	KeyData *keyData;
+	ThreadData *threadSpecData = (ThreadData *)k_malloc(sizeof(ThreadData));
+	sys_dnode_t *nodeL = (sys_dnode_t *)k_malloc(sizeof(sys_dnode_t));
+	int found = 0;
+
+	/* Traverse the list of keys set by the thread, looking for key.
+	 * If the key is already in the list, re-assign its value.
+	 * Else add the key to the thread's list.
+	 */
+	sem_wait(&sem);
+
+	sys_dlist_init(&thread->keyList);
+
+	nodeL = sys_dlist_peek_head(&(thread->keyList));
+
+	if (nodeL != NULL) {
```

Comment:
There are macros to traverse dlists; please use them instead of using open-coded loops such as these.

Reply:
Yeah, this is quite messy I agree, I have replaced these four line conditions in the `if` and `do-while` with a single line macro `SYS_DLIST_FOR_EACH_NODE_SAFE` to loop over the list elements. The test passes with all these modifications, do take a look to see if there's something I have missed.